M2/cmake/build-libraries.cmake

```diff
@@ -0,0 +1,1172 @@
+                      SHARED=$<IF:$<BOOL:${BUILD_SHARED_LIBS}>,on,off>
+                      NTL_STD_CXX14=on
+                      NTL_NO_INIT_TRANS=on # TODO: still necessary?
+                      CPPFLAGS=${CPPFLAGS} # TODO: add -DDEBUG if DEBUG
+                      CXXFLAGS=${CXXFLAGS}
+                      LDFLAGS=${LDFLAGS}
+                      CXX=${CMAKE_CXX_COMPILER}
+                      RANLIB=${CMAKE_RANLIB}
+  COMMAND           ${MAKE} -j${PARALLEL_JOBS}
+  WORKING_DIRECTORY libraries/ntl/build/src
+  EXCLUDE_FROM_MAIN ON
+  USES_TERMINAL ON
+  )
+if(AUTOTUNE)
+  add_dependencies(build-ntl-install build-ntl-wizard)
+endif()
+_ADD_COMPONENT_DEPENDENCY(libraries ntl mp NTL_FOUND)
+
+
+# https://github.com/Macaulay2/flint2
+ExternalProject_Add(build-flint
+  PREFIX            libraries/flint2
+  SOURCE_DIR        ${CMAKE_SOURCE_DIR}/submodules/flint2
+  BINARY_DIR        libraries/flint2/build
+  CMAKE_ARGS        -DCMAKE_INSTALL_PREFIX=${M2_HOST_PREFIX}
+                    -DCMAKE_SYSTEM_PREFIX_PATH=${M2_HOST_PREFIX}
+                    -DCMAKE_MODULE_PATH=${CMAKE_SOURCE_DIR}/cmake
+                    -DCMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE}
+                    -DCMAKE_C_COMPILER=${CMAKE_C_COMPILER}
+                    -DCMAKE_CXX_COMPILER=${CMAKE_CXX_COMPILER}
+                    -DCMAKE_C_FLAGS=${CFLAGS}
+                    -DCMAKE_CXX_FLAGS=${CXXFLAGS}
+                    -DBUILD_SHARED_LIBS=${BUILD_SHARED_LIBS}
+                    -DIPO_SUPPORTED=OFF # TODO: because of clang; see https://github.com/wbhart/flint2/issues/644
+                    -DHAVE_TLS=OFF
+                    -DWITH_NTL=ON
+                    -DWITH_MPIR=${USING_MPIR}
+                    # TODO: force SIMD flags off for distribution
+                    #-DHAS_FLAG_MPOPCNT
+                    #-DHAS_FLAG_UNROLL_LOOPS
+  INSTALL_COMMAND   ${CMAKE_COMMAND} --install . ${strip_setting}
+          COMMAND   ${CMAKE_COMMAND} -E make_directory ${M2_INSTALL_LICENSESDIR}/flint2
+          COMMAND   ${CMAKE_COMMAND} -E copy_if_different ${CMAKE_SOURCE_DIR}/submodules/flint2/README ${M2_INSTALL_LICENSESDIR}/flint2
+          COMMAND   ${CMAKE_COMMAND} -E copy_if_different ${CMAKE_SOURCE_DIR}/submodules/flint2/LICENSE ${M2_INSTALL_LICENSESDIR}/flint2
+  TEST_COMMAND      ${CMAKE_COMMAND} . -DBUILD_TESTING=ON
+       COMMAND      ${CMAKE_COMMAND} --build .
+       COMMAND      ${CMAKE_COMMAND} --build . --target test
+  EXCLUDE_FROM_ALL  ON
+  TEST_EXCLUDE_FROM_MAIN ON
+  STEP_TARGETS      install test
+  )
+_ADD_COMPONENT_DEPENDENCY(libraries flint "mp;mpfr;ntl" FLINT_FOUND)
+
+
+# https://github.com/Singular/Sources/tree/spielwiese/factory
+# TODO: what is ftmpl_inst.o?
+set(factory_NTL_HOME_PATH "${M2_HOST_PREFIX} ${NTL_INCLUDE_DIR}/..")
+set(factory_FLINT_HOME_PATH "${M2_HOST_PREFIX} ${FLINT_INCLUDE_DIR}/..")
+ExternalProject_Add(build-factory
+  GIT_REPOSITORY    ${CMAKE_SOURCE_DIR}/submodules/Singular/.git
+  GIT_TAG           1e4e3f39
```

Comment:
What's this git commit number doing here?  Does it override the commit number recorded by git for the submodule?  If so, that seems to be a recipe for causing a discrepancy between the cmake build and the autoconf build, and probably could be fixed by simply deleting the for git_tag lines in this file.  (Mike noticed this.  @mikestillman )

Reply:
@DanGrayson Currently, Factory, Givaro, and fflas-ffpack dont support out of tree builds, so the solution here has been to use the submodule *as a remote* and make a new, clean clone in the build directory, then build directly there. I'll replace the commit number with HEAD, which I believe would then use the checked out commit from from the submodule.

Reply:
Frobby, too, right?

Reply:
Yes, I forgot that. I replaced all of them with HEAD. Also, I removed the Singular submodule, since the tarfile for 4.1.3 apparently doesn't include the extra `factory-4.1.3` directory, so it is definitely simpler. I also added my patch for it, so you can use that with autotools as well.